src/ImageSharp/Formats/Gif/LzwEncoder.cs

```diff
@@ -11,411 +11,219 @@ namespace SixLabors.ImageSharp.Formats.Gif;
-    /// an adaptive reset, whereby the code table is cleared when the compression
-    /// ratio decreases, but after the table fills.  The variable-length output
-    /// codes are re-sized at this point, and a special CLEAR code is generated
-    /// for the decompressor.  Late addition:  construct the table according to
-    /// file size for noticeable speed improvement on small files.  Please direct
-    /// questions about this implementation to ames!jaw.
-    /// </summary>
-    private int globalInitialBits;
-
-    /// <summary>
-    /// The clear code.
-    /// </summary>
-    private int clearCode;
-
-    /// <summary>
-    /// The end-of-file code.
-    /// </summary>
-    private int eofCode;
-
-    /// <summary>
-    /// Output the given code.
-    /// Inputs:
-    ///      code:   A bitCount-bit integer.  If == -1, then EOF.  This assumes
-    ///              that bitCount =&lt; wordsize - 1.
-    /// Outputs:
-    ///      Outputs code to the file.
-    /// Assumptions:
-    ///      Chars are 8 bits long.
-    /// Algorithm:
-    ///      Maintain a BITS character long buffer (so that 8 codes will
-    /// fit in it exactly).  Use the VAX insv instruction to insert each
-    /// code in turn.  When the buffer fills up empty it and start over.
-    /// </summary>
-    private int currentAccumulator;
-
-    /// <summary>
-    /// The current bits.
-    /// </summary>
-    private int currentBits;
-
-    /// <summary>
-    /// Number of characters so far in this 'packet'
-    /// </summary>
-    private int accumulatorCount;
-
-    /// <summary>
-    /// Initializes a new instance of the <see cref="LzwEncoder"/> class.
-    /// </summary>
-    /// <param name="memoryAllocator">The <see cref="MemoryAllocator"/> to use for buffer allocations.</param>
-    /// <param name="colorDepth">The color depth in bits.</param>
+    private readonly int minCodeSize;
+    private readonly int clearCode;
+    private readonly int eoiCode;
+    private readonly int codeMask;
+    private int nextCode;
+    private int curCodeSize;
+    private int curShift;
+    private int cur;
+    private nint position;
+    private readonly MemoryAllocator memoryAllocator;
+    private readonly IMemoryOwner<byte> bufferOwner;
+    private readonly Span<byte> buffer;
+    private const int MaxStackSize = 1 << 12;
+    private const int SubBlockLength = 255;
+
     public LzwEncoder(MemoryAllocator memoryAllocator, int colorDepth)
     {
-        this.initialCodeSize = Math.Max(2, colorDepth);
-        this.hashTable = memoryAllocator.Allocate<int>(HashSize, AllocationOptions.Clean);
-        this.codeTable = memoryAllocator.Allocate<int>(HashSize, AllocationOptions.Clean);
+        this.memoryAllocator = memoryAllocator;
+        this.minCodeSize = Math.Max(2, colorDepth);
+        this.clearCode = 1 << this.minCodeSize;
+        this.codeMask = this.clearCode - 1;
+        this.eoiCode = this.clearCode + 1;
```

Comment:
Instead of having this a fields, that are initialized, they could be properties too.
* shrinks the size of this (huge) struct by 8 bytes
* the values are very use to compute and JIT will inline these for sure

Reply:
I wasn't actually aware that this made a difference. Do you know of any docs/blogs that chat about it?

I'll move most of them to the `Compress` method body and convert the allocator and min code size to properties.

I absolutely love it when you jump in and review like this. It's always a great learning experience!

Reply:
Thanks ❤️ 

Unfortunately I don't have docs/blogs-links for you. I collected these infos by measuring, reading GH-issues, etc. So very spread around but some things I remember (sometimes 😉).